### PR TITLE
add bye command

### PR DESCRIPTION
**Pull request by Arefsa78**

### CODE REVIEW

File: server.py [hunk 1]
```diff
@@ -54,7 +55,19 @@ def SendInitMessage(self, request, context):
...
         return pb2.InitMessageFromServer()
+    
+    def SendByeCommand(self, request, context):
+        print("Bye command received")
+        print(f'{"#"*20} n={self.number_of_connections} {"#"*20}')
+        with self.lock:
+            self.number_of_connections -= 1
+        if self.number_of_connections <= 0:
+            os._exit(0)
+        return pb2.Empty()
```
Comment (naderzare) [inline server.py:71]: is it possible to return the message and exit?
maybe the player waits for the response
maybe you can add a property isRunning and change it to false, then use it to terminate the server

Comment (naderzare) [reply server.py:71]: import grpc
from concurrent import futures
import time
import pb2_grpc
import pb2  # Import the generated protobuf code

class Game(pb2_grpc.GameServicer):
    def __init__(self):
        self.isRunning = True

    def SendByeCommand(self, request, context):
        self.isRunning = False
        return pb2.Empty()  # Return the response indicating the server will shut down

def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    game_service = Game()
    pb2_grpc.add_GameServicer_to_server(game_service, server)
    server.add_insecure_port('[::]:50051')
    server.start()
    print("Server started at port 50051")
    
    try:
        while game_service.isRunning:
            time.sleep(1)  # Sleeps to reduce CPU usage, checks the flag every 1 second
    except KeyboardInterrupt:
        pass
    finally:
        print("Shutting down server...")
        server.stop(0)  # Gracefully stop the server
        print("Server shutdown complete.")

if __name__ == '__main__':
    serve()

Comment (naderzare) [reply server.py:71]: or

Comment (naderzare) [reply server.py:71]: class Game(pb2_grpc.GameServicer):
    def SendByeCommand(self, request, context):
        def shutdown_server():
            print("Shutting down server...")
            time.sleep(1)  # Sleep for a short duration to ensure response is sent
            server.stop(0)  # Gracefully stop the server

        # Start the shutdown process in a new thread so it does not block the response
        threading.Thread(target=shutdown_server).start()
        return pb2.Empty()  # Return the response immediately

def serve():
    global server
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=22))
    pb2_grpc.add_GameServicer_to_server(Game(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    print("Server started at port 50051")
    server.wait_for_termination()

if __name__ == '__main__':
    serve()

**Review by naderzare — CHANGES_REQUESTED** [review]

File: server.py (hunk 1 above)
Comment (Arefsa78) [reply server.py:71]: done

**Pull request merged**